readthedocs/donate/models.py

```diff
@@ -50,9 +65,56 @@ def __str__(self):
 
     def as_dict(self):
         "A dict respresentation of this for JSON encoding"
+        hash = get_random_string()
+        image_url = reverse(
+            'donate_view_proxy',
+            kwargs={'promo_id': self.pk, 'hash': hash}
+        )
+        # TODO: Store this hash and confirm that a proper hash was sent later
+        link_url = reverse(
+            'donate_click_proxy',
+            kwargs={'promo_id': self.pk, 'hash': hash}
+        )
         return {
             'id': self.analytics_id,
             'text': self.text,
-            'link': self.link,
-            'image': self.image,
+            'link': link_url,
+            'image': image_url,
+            'hash': hash,
         }
+
+    def cache_key(self, type, hash):
+        assert type in IMPRESSION_TYPES
+        return 'promo:{id}:{hash}:{type}'.format(id=self.analytics_id, hash=hash, type=type)
+
+    def incr(self, type):
+        """Add to the number of times this action has been performed, stored in the DB"""
+        assert type in IMPRESSION_TYPES
+        day = get_ad_day()
+        impression, _ = self.impressions.get_or_create(date=day)
+        setattr(impression, type, models.F(type) + 1)
+        impression.save()
+
+        # TODO: Support redis, more info on this PR
+        # github.com/rtfd/readthedocs.org/pull/2105/files/1b5f8568ae0a7760f7247149bcff481efc000f32#r58253051
+
+    def shown(self, day=None):
+        """Return the percentage of times this ad was shown when offered."""
+        if not day:
+            day = get_ad_day()
+        impression = self.impressions.get(date=day)
+        return float(impression.views) / float(impression.offers) * 100.0
```

Comment:
Seems display in the admin is the only case we'll want human readable. We'll be using this as a float for all probability calculations and compounding estimates when we go to run an ad for a full month at a fraction of the views.

Overriding the field display on the model should just require a `def shown(...)` on the admin class